### PR TITLE
counter

### PR DESCRIPTION
**Pull request by jasoncolburne**
## Rationale
This was next in the list of things that could be implemented. It had very few dependencies.

## Changes
- adds `Counter`, `counter::Codex` and associated machinery
- adds `nab_sextets()` to `core::util`
- improves test coverage

## Testing
Extensive testing on `Counter`, as most tests were ported. There were many.

## Notes
Sorry about the length, I got a bit carried away and don't want to ask for review over weekends.

### CODE REVIEW

File: src/core/counter/tables.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+use crate::error::{Error, Result};
+
+pub(crate) fn sizage(s: &str) -> Result<Sizage> {
+    Ok(match s {
+        "-A" => Sizage {
+            hs: 2,
+            ss: 2,
+            fs: 4,
+            ls: 0,
+        },
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:11]: This was way nicer before cargo fmt.. My reasoning was that I wanted to see the labels in the code.

File: src/core/counter/tables.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+        _ => Err(Box::new(Error::UnknownHardage(format!(
+            "{{\
+                \"message\":\"unknown hardage\",\
+                \"code\":{s}\
+            }}"
+        )))),
+    }
+}
+
+pub(crate) fn bardage(b: &[u8]) -> Result<u32> {
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:122]: We should think about computing these for other hardages.

File: src/core/counter/tables.rs [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+    #[test]
+    fn test_sizage() {
+        let mut s: Sizage;
+
+        s = sizage("-A").unwrap();
+        assert_eq!(s.hs, 2);
+        assert_eq!(s.ss, 2);
+        assert_eq!(s.fs, 4);
+        assert_eq!(s.ls, 0);
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:238]: Here is how I generated this (irb/ruby):

```
3.0.0 :001 > s.lines.each do |line|
3.0.0 :002 >   line =~ /\s+'(.+)': Sizage\(hs=(\d), ss=(\d), fs=(\d), ls=(\d)\)/
3.0.0 :003 >   puts "s = sizage(\"#{$1}\");"
3.0.0 :004 >   puts "assert_eq!(s.hs, #{$2});"
3.0.0 :005 >   puts "assert_eq!(s.ss, #{$3});"
3.0.0 :006 >   puts "assert_eq!(s.fs, #{$4});"
3.0.0 :007 >   puts "assert_eq!(s.ls, 0);"
3.0.0 :008 >   puts
3.0.0 :009 > end
```

Comment (jasoncolburne) [reply src/core/counter/tables.rs:238]: from this input: https://github.com/WebOfTrust/keripy/blob/development/src/keri/core/coring.py#L4282-L4296

File: src/core/counter/tables.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        assert_eq!(Codex::TransIdxSigGroups.code(), "-F");
+        assert_eq!(Codex::SealSourceCouples.code(), "-G");
+        assert_eq!(Codex::TransLastIdxSigGroups.code(), "-H");
+        assert_eq!(Codex::SealSourceTriples.code(), "-I");
+        assert_eq!(Codex::SadPathSig.code(), "-J");
+        assert_eq!(Codex::SadPathSigGroup.code(), "-K");
+        assert_eq!(Codex::PathedMaterialQuadlets.code(), "-L");
+        assert_eq!(Codex::AttachedMaterialQuadlets.code(), "-V");
+        assert_eq!(Codex::BigAttachedMaterialQuadlets.code(), "-0V");
+        assert_eq!(Codex::KERIProtocolStack.code(), "--AAA");
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:341]: similarly,

```
3.0.0 :001 > s.lines.each do |line|
3.0.0 :002 >   line =~ /\s+(\w+): str = '(.+)'  # .*$/
3.0.0 :003 >   puts "assert_eq!(Codex::#{$1}.code(), \"#{$2}\");"
3.0.0 :004 > end
```

from https://github.com/WebOfTrust/keripy/blob/development/src/keri/core/coring.py#L4165-L4179

File: src/core/counter/tables.rs [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    TransIdxSigGroups,
+    SealSourceCouples,
+    TransLastIdxSigGroups,
+    SealSourceTriples,
+    SadPathSig,
+    SadPathSigGroup,
+    PathedMaterialQuadlets,
+    AttachedMaterialQuadlets,
+    BigAttachedMaterialQuadlets,
+    KERIProtocolStack,
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:164]: this one was a simple transformation

File: src/core/counter/tables.rs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+impl Codex {
+    pub(crate) fn code(&self) -> &'static str {
```
Comment (jasoncolburne) [inline src/core/counter/tables.rs:168]: these ones are a more complex, examples of how I did this posted below

File: .rustfmt.toml [hunk 7]
```diff
@@ -0,0 +1 @@
+use_small_heuristics = "Max"
```
Comment (jasoncolburne) [inline .rustfmt.toml:1]: This made the code more compact, the sizages were blown out over multiple lines otherwise.

File: src/error.rs [hunk 8]
```diff
@@ -0,0 +1,195 @@
+pub type BoxedError = Box<dyn std::error::Error>;
+pub type Result<T> = core::result::Result<T, BoxedError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(
+        "{{\
+                \"variant\":\"Matter\",\
+                \"meta\":{0}\
+            }}"
```
Comment (jasoncolburne) [inline src/error.rs:10]: Sigh. I see something broke this formatting. I'll fix it later.

File: src/core/matter/tables.rs [hunk 9]
```diff
@@ -1,79 +1,66 @@
...
+        "8AAA" => Sizage { hs: 4, ss: 4, fs: 0, ls: 1 },
+        "9AAA" => Sizage { hs: 4, ss: 4, fs: 0, ls: 2 },
+        "4B" => Sizage { hs: 2, ss: 2, fs: 0, ls: 0 },
+        "5B" => Sizage { hs: 2, ss: 2, fs: 0, ls: 1 },
+        "6B" => Sizage { hs: 2, ss: 2, fs: 0, ls: 2 },
+        "7AAB" => Sizage { hs: 4, ss: 4, fs: 0, ls: 0 },
+        "8AAB" => Sizage { hs: 4, ss: 4, fs: 0, ls: 1 },
+        "9AAB" => Sizage { hs: 4, ss: 4, fs: 0, ls: 2 },
+        _ => {
+            return Err(Box::new(Error::UnknownSizage(s.to_string())));
```
Comment (jasoncolburne) [inline src/core/matter/tables.rs:53]: Looks like I missed converting some couple error strings.

File: src/core/matter/tables.rs [hunk 10]
```diff
@@ -72,10 +66,23 @@ pub(crate) fn hardage(c: char) -> Result<i32> {
...
+            }}"
+            .to_string(),
         ))),
-        '_' => Err(Box::new(Error::UnexpectedCode("op code start".to_string()))),
-        _ => Err(Box::new(Error::UnknownHardage(c.to_string()))),
+        _ => Err(Box::new(Error::UnknownHardage(format!(
+            "{{\
+                \"message\":\"unknown hardage\",\
+                \"hard\":{c}\
+            }}"
```
Comment (jasoncolburne) [inline src/core/matter/tables.rs:84]: This is really fragile and the json is unvalidated. I'm going to be pushing a change that uses `serde_json::json!` to make things much easier to maintain

Comment (jasoncolburne) [reply src/core/matter/tables.rs:84]: Not to mention the gross escaping.

File: src/error.rs [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+pub type Result<T> = core::result::Result<T, BoxedError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(
+        "{{\
+            \"variant\":\"Matter\",\
+            \"details\":{0}\
+        }}"
+    )]
```
Comment (jasoncolburne) [inline src/error.rs:11]: The `json!` macro doesn't work here 😞 so this is a bit fragile. I'll add tests for these errors perhaps, that try to parse an instantiated error's json output string using `serde_json` or `json`.

Comment (jasoncolburne) [reply src/error.rs:11]: Tests implemented.

**Comment by jasoncolburne** [comment]
Errors now occupy two lines of code rather than one. Test coverage actually improved in this PR. Here's what I'm talking about:

Screenshot 2023-01-29 at 12 36 58 PM

File: src/core/matter/mod.rs [hunk 12]
```diff
@@ -48,63 +51,86 @@ impl Matter {
...
+            let cs = szg.hs + szg.ss;
+            ((szg.fs - cs) * 3 / 4) - szg.ls
         };
 
-        Ok(Matter {
-            code,
-            size,
-            raw: raw[..rize as usize].to_owned(),
-        })
+        if raw.len() < rize as usize {
```
Comment (jasoncolburne) [inline src/core/matter/mod.rs:102]: adding sad path tests caught this deficiency! 👍🏻

File: src/core/counter/tables.rs (hunk 2 above)
Comment (m00sey) [reply src/core/counter/tables.rs:35]: yeah that's a great suggestion. Can you open an issue?

File: src/core/diger.rs [hunk 13]
```diff
@@ -427,7 +407,7 @@ mod test_diger {
     }
 
     #[test]
-    fn test_python_parity() {
+    fn test_python_interop() {
```
Comment (m00sey) [inline src/core/diger.rs:410]: :sparkling_heart:

File: src/core/matter/mod.rs [hunk 14]
```diff
@@ -136,56 +139,61 @@ impl Matter {
     fn infil(&self) -> Result<String> {
         let code = &self.code;
         let size = self.size;
-        let mut raw = self.raw.to_vec();
+        let mut raw = self.raw.clone();
 
         let ps = (3 - raw.len() % 3) % 3;
-        let sizage_ = tables::sizage(code)?;
+        let szg = tables::sizage(code)?;
```
Comment (m00sey) [inline src/core/matter/mod.rs:145]: ha! yeah ok that is better than sizage_

File: src/core/matter/mod.rs [hunk 15]
```diff
@@ -387,58 +398,63 @@ impl Matter {
                     "unexpected start during extraction".to_string(),
                 )));
             } else {
+                // unreachable
```
Comment (m00sey) [inline src/core/matter/mod.rs:401]: yeah I think for now this is better an the panic of unreachable()

File: src/core/sizage.rs [hunk 16]
```diff
@@ -0,0 +1,7 @@
+#[derive(Debug, PartialEq)]
+pub struct Sizage {
```
Comment (m00sey) [inline src/core/sizage.rs:2]: We will need to move this under matter, as Indexer also has Sizage (Xizage) in python

**Review by m00sey — APPROVED** [review]

**Pull request merged**